src/invite-api.js

```diff
@@ -1,236 +1,440 @@
 // @ts-check
 import { TypedEmitter } from 'tiny-typed-emitter'
 import { InviteResponse_Decision } from './generated/rpc.js'
-import { projectKeyToId, projectKeyToPublicId } from './utils.js'
+import { assert, keyToId } from './utils.js'
+import timingSafeEqual from './lib/timing-safe-equal.js'
+import { promiseWithResolvers } from './ponyfills.js'
 
 /**
- * @typedef {Object} InviteApiEvents
+ * Internally, we typically use the `Invite` type from the protobuf. We also use
+ * an external type for public consumers.
  *
- * @property {(info: { projectId: string, projectName?: string, peerId: string, roleName: string, roleDescription: string | undefined, invitorName: string }) => void} invite-received
+ * @internal
+ * @typedef {import('./generated/rpc.js').Invite} InviteInternal
+ */
+
+/**
+ * @internal
+ * @typedef {import('./generated/rpc.js').ProjectJoinDetails} ProjectJoinDetails
  */
 
 /**
- * @template K
- * @template V
- * @extends {Map<K, Set<V>>}
+ * @typedef {object} Invite
+ * @prop {string} inviteId
+ * @prop {string} projectName
+ * @prop {string} [roleName]
+ * @prop {string} [roleDescription]
+ * @prop {string} [invitorName]
  */
-class MapOfSets extends Map {
-  /** @param {K} key */
-  get(key) {
-    const existing = super.get(key)
-    if (existing) return existing
-    /** @type {Set<V>} */
-    const set = new Set()
-    super.set(key, set)
-    return set
+
+/**
+ * Manage pending invite state.
+ */
+class PendingInvites {
+  /**
+   * @internal
+   * @typedef {object} PendingInvite
+   * @prop {string} peerId
+   * @prop {InviteInternal} invite
+   * @prop {boolean} isAccepting
+   */
+
+  /** @type {Map<string, PendingInvite>} */
+  #byInviteId = new Map()
+
+  /** @type {Map<string, PendingInvite[]>} */
+  #byProjectPublicId = new Map()
```

Comment:
Having this map feels like a premature optimization - I think it's just an index of `#byInviteId`, but having it requires a lot of additional code to keep it in sync with `#byInviteId`. Is it necessary?

Reply:
Good point. Removed in 539916f976eb3927cd1b955e8b1fcaca20233f87.